cmd/openshift-tuned.go

```diff
@@ -238,27 +246,162 @@ func apiActiveProfile(w http.ResponseWriter, req *http.Request) {
+		// Pod labels changed
+		if tuned.podLabels == nil {
+			tuned.podLabels = map[string]map[string]string{}
+		}
+		tuned.podLabels[key] = pod.Labels
+		tuned.change.pod = true
+		return
+	}
+
+	// Pod labels didn't change, event didn't modify labels
+}
+
+func eventWatch(w watch.Interface, f func(watch.Event, *tunedState), tuned *tunedState) {
+	defer w.Stop()
+	for event := range w.ResultChan() {
+		f(event, tuned)
+	}
+}
+
+func timedTunedReloader(tuned *tunedState) {
+	ticker := time.NewTicker(time.Second * time.Duration(labelDumpInterval))
+
+	for range ticker.C {
+		// Check pod labels
+		var reload bool = false
+		if tuned.change.pod {
+			// Pod labels changed
+			tuned.change.pod = false
+			podLabelsDump(tuned.podLabels, filePodLabels)
+			reload = true
+		}
+		// Check node labels
+		if tuned.change.node {
+			// Node labels changed
+			tuned.change.node = false
+			nodeLabelsDump(tuned.nodeLabels, fileNodeLabels)
+			reload = true
+		}
+		// Check tuned profiles/recommend file changes
+		if tuned.change.cfg {
+			tuned.change.cfg = false
+			profilesExtract()
+			reload = true
+		}
+		if reload {
+			tunedReload()
+		}
+	}
+}
+
+func pullLabels(clientset *kubernetes.Clientset, tuned *tunedState, nodeName string) {
+	var (
+		err                  error
+		resyncPeriodDuration int64 = resyncPeriodDefault
+	)
+
+	if os.Getenv("RESYNC_PERIOD") != "" {
+		resyncPeriodDuration, err = strconv.ParseInt(os.Getenv("RESYNC_PERIOD"), 10, 64)
+		if err != nil {
+			log.Printf("Error: cannot parse RESYNC_PERIOD (%s), using %d\n", os.Getenv("RESYNC_PERIOD"), resyncPeriodDefault)
+			resyncPeriodDuration = resyncPeriodDefault
+		}
+	}
+
+	ticker := time.NewTicker(time.Second * time.Duration(resyncPeriodDuration))
+
+	for range ticker.C {
+		// Resync period elapsed, force-pull node and pod labels
+		nodeLabels := nodeLabelsGet(clientset, nodeName)
+		if !reflect.DeepEqual(nodeLabels, tuned.nodeLabels) {
+			tuned.nodeLabels = nodeLabels
+			tuned.change.node = true
+		}
+		podLabels := podLabelsGet(clientset, nodeName)
+		if !reflect.DeepEqual(podLabels, tuned.podLabels) {
+			tuned.podLabels = podLabels
+			tuned.change.pod = true
+		}
+	}
+}
+
+func mainLoop(clientset *kubernetes.Clientset, nodeName string) {
+	var tuned tunedState
 
 	if *apiPort > 0 {
 		go func() {
 			mux := http.NewServeMux()
 			mux.HandleFunc("/active_profile", apiActiveProfile)
-			log.Printf("%s: listening on %d\n", PNAME, *apiPort)
+			log.Printf("Listening on %d\n", *apiPort)
 			log.Fatal(http.ListenAndServe(fmt.Sprintf((":%d"), *apiPort), mux))
 		}()
 	}
 
-	ticker := time.NewTicker(time.Second * time.Duration(resyncPeriodDuration))
-	go func() {
-		for range ticker.C {
-			nodeLabels := nodeLabelsGet(clientset, nodeName)
-			nodeLabelsCompare(nodeLabelsOld, nodeLabels)
-			nodeLabelsOld = nodeLabels
-		}
-	}()
+	// Watch for node and pod label changes
+	if *boolWatchLabels {
+		go eventWatch(nodeWatch(clientset, nodeName), nodeChangeHandler, &tuned)
```

Comment:
These goroutines seem stuck in endless loops?

Reply:
openshift-tuned is an "init-like" process, so yes, it runs "forever".

Reply:
Yeah the problem would be if the binary ever terminates, not sure if we're concerned with that?

Reply:
Can you expand on the problems you're seeing?

Reply:
You're spinning off several goroutines that will never stop running so wouldn't that be a goroutine leak?

Reply:
Notice that I *asked* if it was a goroutine leak? By my estimation having goroutines that continue running after program termination would qualify. If that's *intended* and there are no other resources won't be automatically cleaned up (that don't need to be closed manually) in those goroutines then I guess it would be okay to keep. Unfortunately I do not have all the answers for everything @jmencak, maybe one day.

Reply:
Based on our discussion, we will leave the goroutines as they are for now. If we figure out what to do with them in the future, we'll revisit this.